### PR TITLE
fix(streaming): fix reporting & stopping of rearranged chain

### LINKED ISSUE #1757: storage: epoch of table iterator for batch query executor is not pinned
**Describe the bug**
We use several range scans in the table iterator to avoid pinning the initial version for a long time. Note that every range scan may pin a different version, while we should ensure that the data in the given epoch always exist. Thus, we must assume the epoch is pinned.

There're two cases where the table iterator (`CellBasedTableIter`) is used. One for ad-hoc batch queries, where the frontend pins the epoch during the execution. The other is for the `BatchQueryExecutor` while creating MV on MV, which is not pinned and might be problematic.

It seems that the compaction is triggered by the transaction commit. Before, the process of creating MVs always finished in a single epoch, so it's hard to reproduce this bug. However, when testing `RearrangedChain`, which makes the creation last for several epochs, we may encounter an "expired epoch" error.

```
2022-04-11T05:22:16.110176Z  WARN risingwave_stream::executor::actor: Actor polling failed: Stream error: executor v1 error Stream error: executor v1 error Stream error: storage error Hummock error: Expired Epoch: watermark 108111759615787008, epoch 108111759573188608.
  backtrace of inner error:
   0: std::backtrace_rs::backtrace::libunwind::trace
             at /rustc/1eb72580d076935a3e590deb6e5813a5aef3eca4/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1: std::backtrace_rs::backtrace::trace_unsynchronized
             at /rustc/1eb72580d076935a3e590deb6e5813a5aef3eca4/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2: std::backtrace::Backtrace::create
             at /rustc/1eb72580d076935a3e590deb6e5813a5aef3eca4/library/std/src/backtrace.rs:328:13
   3: <risingwave_storage::hummock::error::HummockError as core::convert::From<risingwave_storage::hummock::error::HummockErrorInner>>::from
             at ./src/storage/src/hummock/error.rs:54:10
   4: <T as core::convert::Into<U>>::into
             at /rustc/1eb72580d076935a3e590deb6e5813a5aef3eca4/library/core/src/convert/mod.rs:546:9
   5: risingwave_storage::hummock::error::HummockError::expired_epoch
             at ./src/storage/src/hummock/error.rs:108:9
   6: risingwave_storage::hummock::utils::validate_epoch
             at ./src/storage/src/hummock/utils.rs:58:20
   7: <risingwave_storage::hummock::HummockStorage as risingwave_storage::store::StateStore>::iter::{{closure}}
             at ./src/storage/src/hummock/mod.rs:421:13
   8: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/1eb72580d076935a3e590deb6e5813a5aef3eca4/library/core/src/future/mod.rs:91:19
   9: <risingwave_storage::hummock::HummockStorage as risingwave_storage::store::StateStore>::scan::{{closure}}
             at ./src/storage/src/hummock/mod.rs:316:55
  10: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/1eb72580d076935a3e590deb6e5813a5aef3eca4/library/core/src/future/mod.rs:91:19
  11: <risingwave_storage::monitor::monitored_store::MonitoredStateStore<S> as risingwave_storage::store::StateStore>::scan::{{closure}}
             at ./src/storage/src/monitor/monitored_store.rs:100:66
  12: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/1eb72580d076935a3e590deb6e5813a5aef3eca4/library/core/src/future/mod.rs:91:19
  13: risingwave_storage::keyspace::Keyspace<S>::scan_with_start_key::{{closure}}
             at ./src/storage/src/keyspace.rs:143:45
  14: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/1eb72580d076935a3e590deb6e5813a5aef3eca4/library/core/src/future/mod.rs:91:19
  15: risingwave_storage::table::cell_based_table::CellBasedTableRowIter<S>::consume_more::{{closure}}
             at ./src/storage/src/table/cell_based_table.rs:326:92
  16: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/1eb72580d076935a3e590deb6e5813a5aef3eca4/library/core/src/future/mod.rs:91:19
  17: <risingwave_storage::table::cell_based_table::CellBasedTableRowIter<S> as risingwave_storage::table::TableIter>::next::{{closure}}
             at ./src/storage/src/table/cell_based_table.rs:394:40
  18: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/1eb72580d076935a3e590deb6e5813a5aef3eca4/library/core/src/future/mod.rs:91:19
  19: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/1eb72580d076935a3e590deb6e5813a5aef3eca4/library/core/src/future/future.rs:124:9
  20: risingwave_storage::table::cell_based_table::CellBasedTableRowIter<S>::collect_data_chunk::{{closure}}
             at ./src/storage/src/table/cell_based_table.rs:350:30
  21: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/1eb72580d076935a3e590deb6e5813a5aef3eca4/library/core/src/future/mod.rs:91:19
  22: risingwave_stream::executor_v2::batch_query::BatchQueryExecutor<S>::execute_inner::{{closure}}
             at ./src/stream/src/executor_v2/batch_query.rs:52:14
```

**To Reproduce**
Creating MVs with the `RearrangedChain` executor may lead to this bug.

cc @hzxa21 @zwang28

### PR DESCRIPTION
**Pull request by BugenZhao**
## What's changed and what's your intention?

We should report success right before the first barrier after we stopped the rearrangement, or the user may find stale data in an epoch. This might be the cause of e2e failure of https://github.com/singularity-data/risingwave/runs/5787805327.

Besides, it's incorrect to check whether to stop using `try_recv` in the loop, which may cause dead locks. This PR use `stream::select` to resolve this.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
- #1528
- #1540
- Fix #1757

### CODE REVIEW

File: src/stream/src/executor_v2/mview/state.rs [hunk 1]
```diff
@@ -66,10 +66,8 @@ impl<S: StateStore> ManagedMViewState<S> {
         assert_eq!(self.order_types.len(), pk.size());
         assert_eq!(self.column_ids.len(), value.size());
 
-        let l = pk.0.len() == 3
-            && pk
-                .0
-                .iter()
+        let l =
+            pk.0.iter()
```
Comment (MrCroxx) [inline src/stream/src/executor_v2/mview/state.rs:70]: Add some comments here, 👀?

File: src/stream/src/executor_v2/mview/state.rs [hunk 2]
```diff
@@ -84,10 +82,8 @@ impl<S: StateStore> ManagedMViewState<S> {
     pub fn delete(&mut self, pk: Row) {
         assert_eq!(self.order_types.len(), pk.size());
 
-        let l = pk.0.len() == 3
-            && pk
-                .0
-                .iter()
+        let l =
+            pk.0.iter()
```
Comment (MrCroxx) [inline src/stream/src/executor_v2/mview/state.rs:86]: Ditto.

File: src/stream/src/executor_v2/mview/state.rs (hunk 1 above)
Comment (BugenZhao) [reply src/stream/src/executor_v2/mview/state.rs:70]: This is for debugging. 😢 I'll remove it when it's ready.

**Comment by BugenZhao** [comment]
@zwang28 PTAL at the Hummock part. Thanks!

**Review by zwang28 — COMMENTED** [review]
LGTM for hummock manager part

**Comment by BugenZhao** [comment]
Seems still unstable. Let's merge the fixes without switching to rearranged chain.

**Pull request merged**